File: back/world/world.go
package world

import (
	"root/common"
	"root/mechanics"
)

type World struct {
	ID       string
	TimeStep float32
	CurTime  float32
	Objects  []*common.Object
}

type Frame struct {
	CurTime        float32
	IDArray        []string
	PositionArrays [][]float32
	VertexArrays   [][]float32
	IndexArrays    [][]uint32
	ColorArray     []string
}

func NewWorld(id string, timeStep float32) *World {
	return &World{
		ID:       id,
		TimeStep: timeStep,
	}
}

func (w *World) Flatten() Frame {
	IDArray := make([]string, 0, len(w.Objects))
	PositionArrays := make([][]float32, 0, len(w.Objects))
	VertexArrays := make([][]float32, 0, len(w.Objects))
	IndexArrays := make([][]uint32, 0, len(w.Objects))
	ColorArray := make([]string, 0, len(w.Objects))

	for _, obj := range w.Objects {
		id, pos, verts, inds, color := obj.Flatten()
		IDArray = append(IDArray, id)
		PositionArrays = append(PositionArrays, pos)
		VertexArrays = append(VertexArrays, verts)
		IndexArrays = append(IndexArrays, inds)
		ColorArray = append(ColorArray, color)
	}

	return Frame{
		CurTime:        w.CurTime,
		IDArray:        IDArray,
		PositionArrays: PositionArrays,
		VertexArrays:   VertexArrays,
		IndexArrays:    IndexArrays,
		ColorArray:     ColorArray,
	}
}

func (w *World) AllPairs() [][2]*common.Object {
	var pairs [][2]*common.Object
	for i := 0; i < len(w.Objects); i++ {
		for j := i + 1; j < len(w.Objects); j++ {
			pairs = append(pairs, [2]*common.Object{w.Objects[i], w.Objects[j]})
		}
	}
	return pairs
}

// Find all pairs of objects whose AABBs overlap
func (w *World) Overlaps() [][2]*common.Object {
	var overlaps [][2]*common.Object
	for i := 0; i < len(w.Objects); i++ {
		a := w.Objects[i]
		for j := i + 1; j < len(w.Objects); j++ {
			b := w.Objects[j]
			if a.TriangleOverlap(b) {
				overlaps = append(overlaps, [2]*common.Object{a, b})
			}
		}
	}
	return overlaps
}

func (w *World) Update() {

	w.CurTime += w.TimeStep

	overlapObjectPairs := w.Overlaps()

	for _, pair := range overlapObjectPairs {
		pair[0].StepBack(w.TimeStep)
		pair[1].StepBack(w.TimeStep)
		mechanics.ElasticCollisionResponse(pair[0], pair[1])
	}

	allPairs := w.AllPairs()

	for _, pair := range allPairs {
		mechanics.UniversalGravitationResponse(pair[0], pair[1])
	}

	for i := range w.Objects {
		w.Objects[i].Update(w.TimeStep)
	}

}

File: back/world.py
"""
Physics world simulation.
"""
from typing import List
from obj import Obj
import config
from collision import check_collision_pair
from mechanics import apply_gravity_batched, apply_collision_response


class World:
    def __init__(self, world_id: str, time_step: float):
        """
        Args:
            world_id: Identifier
            time_step: Simulation timestep (seconds)
        """
        self.id = world_id
        self.dt = time_step
        self.time = 0.0
        self.objects: List[Obj] = []

    def add_object(self, obj: Obj):
        self.objects.append(obj)

    def update(self):
        """Run one simulation step."""
        self.time += self.dt

        # 1. Find collisions (parallel collision detection)
        collisions = []
        for i in range(len(self.objects)):
            for j in range(i + 1, len(self.objects)):
                if check_collision_pair(self.objects[i], self.objects[j]):
                    collisions.append((i, j))

        # 2. Resolve collisions
        for i, j in collisions:
            self.objects[i].step_back(self.dt)
            self.objects[j].step_back(self.dt)
            apply_collision_response(self.objects[i], self.objects[j])

        # 3. Apply forces (BATCHED gravitational computation)
        apply_gravity_batched(self.objects)

        # 4. Update all objects
        for obj in self.objects:
            obj.update(self.dt)

    def get_frame_data(self):
        """
        Get frame data for serialization (compatible with Go backend format).

        Returns:
            dict with frame data
        """
        return {
            'cur_time': self.time,
            'id_array': [obj.id for obj in self.objects],
            'position_arrays': [obj.position.cpu().tolist() for obj in self.objects],
            # (3, 3) rotation matrix flattened to 9 floats (row-major)
            'rotation_arrays': [obj.rotation.cpu().reshape(-1).tolist() for obj in self.objects],
            'vertex_arrays': [obj.mesh.vertices.cpu().reshape(-1).tolist() for obj in self.objects],
            'index_arrays': [obj.mesh.indices.cpu().reshape(-1).tolist() for obj in self.objects],
            'color_array': [obj.color for obj in self.objects],
        }